### DISCUSSION TITLE
Increase cache limit per repository

### DISCUSSION THREAD

**Discussion by smorimoto** [actions/cache#497] [General]
Well, I made a similar request for this over a year ago (https://github.com/actions/cache/issues/6), and as a result it has increased to 5GB, but I should say that for large C++ projects, 5GB is still quite insufficient. I think many developers involved in native development have the similar opinion, what do you think?

**Comment by ad-m** [comment]
GitHub Actions support charging for "Storage for Actions and Packages". 

Would removing the cache size limit and introducing charges for cache space will solve your problem? What cache management features do you need to avoid unexpected costs and other billing issues?

**Reply by smorimoto** [reply]
It's not related to `actions/cache`. There was certainly a plan to have a paid plan in the future, but it hasn't come yet. Also, even if we can increase the limit with a paid plan, it would be a little bit off from this discussion since it's not a good thing to have to charge for an open source project to use more than 5GB.

**Reply by ad-m** [reply]
I am discussing to work out an effective solution. I am able to see that introducing a limit protects GitHub against excessive costs. Someone has to finance the cost of the increased storage usage because nothing is for free.

A valuable remark that shifting costs to users will not solve the challenges of open-source projects. However, open-source support depends on what resources GitHub wants to support open-source projects.

**Reply by smorimoto** [reply]
Yes, I know. So, I didn't say to remove the limit, but I said to increase the limit. You talked about costs, but you're only thinking about storage costs, but increasing the cache limit can reduce the time it takes to run the runners, so we need to find a reasonable limit. And 5GB is too small for now. It would be better to increase it to 10GB and take some statistics.

**Reply by dhadka** [reply]
@smorimoto Thanks for opening this discussion.  I definitely agree that the cache limit should be set in a way to let projects efficiently use time on runners since storage costs are significantly less than VM costs.

For these projects, are you/they trying to create one large cache (> 5 GBs) and hitting the upload limit?  Or are they creating many smaller caches (e.g., ten 1 GB caches) which are being evicted leading to additional build time?  Do you have any examples of public projects hitting these limits?  Would love to take a look at them.

**Reply by smorimoto** [reply]
At least in our case, if you try to test 6 or so versions on 3 images, this is a problem. The size per cache is typically around 500MB. I've edited the OCaml to C++ in the body to make it easier for many people to understand, but OCaml often needs this testing approach. (and 90% of the repositories we need are open-source.)

**Reply by smorimoto** [reply]
Because of this cache limit, most projects haven't migrated to GitHub Actions yet, but we are already preparing to do so, for example, some repositories of the MirageOS project may migrate.

**Reply by dhadka** [reply]
@smorimoto Just a quick update...we're looking at the current cache stats and discussing this issue internally.  I'll update here in a week or two when I know more.

**Reply by smorimoto** [reply]
Very very nice, whatever the answer is, I look forward to it as it will help us choose a service in the future.

**Reply by Ailrun** [reply]
As one more cases, HLS (Haskell Language Server) also hits the cache limit quite often and cannot help but removing some of medium priority tasks...

**Reply by smorimoto** [reply]
Is there any update about this? This is a particular problem these days. setup-ocaml uploads the build cache internally, but requires a minimum size of 500 MB per matrix. Without it, the build time would be several times slower, so it's really important to increase the limit.

**Reply by smorimoto** [reply]
@dhadka ping

**Reply by dhadka** [reply]
@smorimoto We've added this to a proposal for caching v3, which includes other requested features like the option to clear the cache through the UI.  We would most likely double the cap to 10 GBs for public repos.  The main friction right now is getting this approved and prioritized alongside everything else we have going on.

There is a longer-term goal of combining cache storage with an account's overall shared storage, which would provide more control over the lifetime and size of the cache.  https://github.com/github/roadmap/issues/66

**Reply by smorimoto** [reply]
@dhadka Superb, looking forward to it 🙂 (If necessary, feel free to forward me anything and I can try to come up with some arguments.)

**Reply by smorimoto** [reply]
@dhadka Is there any updates on this? The 10 GB limit is really what the OCaml community needs.

**Reply by smorimoto** [reply]
Even if you plan to increase the limit on paid plans, it's really important to increase it to 10GB for most open-source projects.

**Reply by dhadka** [reply]
@ahdbilal Anything you can share on the state of ☝️ ?

**Reply by Ailrun** [reply]
> what the OCaml community needs.

+1 from Haskell community.

**Reply by ahdbilal** [reply]
Thanks for all the feedback. I will prioritize this request and will share an update soon.

**Reply by smorimoto** [reply]
Thank you. I look forward to hearing good news.

**Reply by JanJakes** [reply]
Great! This may be very useful for [Docker image caching](https://github.com/docker/build-push-action/pull/406#issuecomment-879184394) that's just about to land in Docker build-push action.

**Reply by FlorianFranzen** [reply]
In our case, just adding one extra dependency increased the cached assets in a way that our different workflows now evict oneanother from cache increasing the CI from a few minutes to multiple hours. I would prefer to raise the limit instead of splitting the repo, but right now this does not seem to be possible. 

Not to mention that economically, Github Actions now has to deal with extremely long running jobs which probably cost much more resources then the a few more gigabytes of cache. 

Not sure why this is such a big issue to increase this limit, even if it was only on request. The average abuser of free ci and caching would just work around this limit using multiple repos anyway.

@ahdbilal Could you please update us on the current state of this.

**Reply by ahdbilal** [reply]
Hey folks! We are very close to shipping this out. Expect something around the 15th of November!

PS: Dates can slip but the team is doing engineering on it so I am hoping we can get this out soon!

**Reply by smorimoto** [reply]
After all, we don't even know what will be shipped yet, but what will be shipped?

**Reply by ahdbilal** [reply]
@smorimoto I should have clarified it before. We are working to increase the per repo cache limit from 5GB to 10GB. The increased size will be available for free to all accounts.

**Reply by smorimoto** [reply]
That's good! Looking forward to that 🙃

**Reply by smorimoto** [reply]
I'm excited to see https://github.com/actions/toolkit/pull/934!

**Comment by N-Usha** [comment]
Happy to announce that today we shipped [cache size increase from 5GB per repo to 10GB](https://github.blog/changelog/2021-11-23-github-actions-cache-size-is-now-increased-to-10gb-per-repository/).    :rocket: :rocket:

**Reply by smorimoto** [reply]
That's good! Thanks for your efforts on this!

**Comment by samus-aran** [comment]
We're locking this discussion because it has not had recent activity and/or other members have asked for more information to assist you but received no response. Thank you for helping us maintain a productive and tidy community for all our members.
